src/sparse_blas/backends/cusparse/operations/cusparse_spmm.cpp

```diff
@@ -0,0 +1,296 @@
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#include "oneapi/mkl/sparse_blas/detail/cusparse/onemkl_sparse_blas_cusparse.hpp"
+
+#include "sparse_blas/backends/cusparse/cusparse_error.hpp"
+#include "sparse_blas/backends/cusparse/cusparse_helper.hpp"
+#include "sparse_blas/backends/cusparse/cusparse_task.hpp"
+#include "sparse_blas/backends/cusparse/cusparse_handles.hpp"
+#include "sparse_blas/common_op_verification.hpp"
+#include "sparse_blas/macros.hpp"
+#include "sparse_blas/matrix_view_comparison.hpp"
+#include "sparse_blas/sycl_helper.hpp"
+
+namespace oneapi::mkl::sparse {
+
+// Complete the definition of the incomplete type
+struct spmm_descr {
+    detail::generic_container workspace;
+    std::size_t temp_buffer_size = 0;
+    bool buffer_size_called = false;
+    bool optimized_called = false;
+    oneapi::mkl::transpose last_optimized_opA;
+    oneapi::mkl::transpose last_optimized_opB;
+    oneapi::mkl::sparse::matrix_view last_optimized_A_view;
+    oneapi::mkl::sparse::matrix_handle_t last_optimized_A_handle;
+    oneapi::mkl::sparse::dense_matrix_handle_t last_optimized_B_handle;
+    oneapi::mkl::sparse::dense_matrix_handle_t last_optimized_C_handle;
+    oneapi::mkl::sparse::spmm_alg last_optimized_alg;
+};
+
+} // namespace oneapi::mkl::sparse
+
+namespace oneapi::mkl::sparse::cusparse {
+
+void init_spmm_descr(sycl::queue& /*queue*/, spmm_descr_t* p_spmm_descr) {
+    *p_spmm_descr = new spmm_descr();
+}
+
+sycl::event release_spmm_descr(sycl::queue& queue, spmm_descr_t spmm_descr,
+                               const std::vector<sycl::event>& dependencies) {
+    return detail::submit_release(queue, spmm_descr, dependencies);
+}
+
+inline auto get_cuda_spmm_alg(spmm_alg alg) {
+    switch (alg) {
+        case spmm_alg::coo_alg1: return CUSPARSE_SPMM_COO_ALG1;
+        case spmm_alg::coo_alg2: return CUSPARSE_SPMM_COO_ALG2;
+        case spmm_alg::coo_alg3: return CUSPARSE_SPMM_COO_ALG3;
+        case spmm_alg::coo_alg4: return CUSPARSE_SPMM_COO_ALG4;
+        case spmm_alg::csr_alg1: return CUSPARSE_SPMM_CSR_ALG1;
+        case spmm_alg::csr_alg2: return CUSPARSE_SPMM_CSR_ALG2;
+        case spmm_alg::csr_alg3: return CUSPARSE_SPMM_CSR_ALG3;
+        default: return CUSPARSE_SPMM_ALG_DEFAULT;
+    }
+}
+
+inline void fallback_alg_if_needed(oneapi::mkl::sparse::spmm_alg& alg, oneapi::mkl::transpose opA,
+                                   oneapi::mkl::transpose opB) {
+    if (alg == oneapi::mkl::sparse::spmm_alg::csr_alg3 &&
+        (opA != oneapi::mkl::transpose::nontrans || opB == oneapi::mkl::transpose::conjtrans)) {
+        // Avoid warnings printed on std::cerr
```

Comment:
What warnings exactly are we avoiding here?

Reply:
Also, if these are runtime warnings from cuSPARSE, shouldn't we `throw unimplemented` exception instead for this case, unless cuSPARSE explicitly mentions that they will use `default_alg`?

Reply:
The warning is:
```
** On entry to cusparseSpMM_bufferSize(): SPMM_CSR_ALG3 is not supported when opA != OPERATION_NON_TRANSPOSE or opB == OPERATION_CONJUGATE_TRANSPOSE
```

You're right I couldn't find what is the behavior of cuSPARSE exactly in this case. The tests pass so it must fallback to something but I don't know what.
I'm not keen on keeping these warnings as they are printed in our tests multiple times and could raise suspicions.
I have a preference for the current behavior, would it help if we add this in the our documentation? We could throw unimplemented if you prefer. Let me know what you think now that you have more information.

Reply:
[CUDA 12.6/cuSPARSE documentation](https://docs.nvidia.com/cuda/cusparse/#cusparsespmm) for `SPMM_CSR_ALG3` says:
> It supports only opA == CUSPARSE_OPERATION_NON_TRANSPOSE
It does not support opB == CUSPARSE_OPERATION_CONJUGATE_TRANSPOSE

I'm fine with oneMKL Interfaces throwing an `unimplemented` exception and documenting that for `SPMM_CSR_ALG3` case because it is documented behavior from cuSPARSE side. Let's not choose the `default_alg` to be fallback. If cuSPARSE decides in the future that they wany to support it, then we can be enable it in this library.

Reply:
This is done in https://github.com/oneapi-src/oneMKL/pull/527/commits/77d19e8f4da0f1fc7c63d26f54b243831bd295f3